File: storelib/src/lib/store/Idx.java
package ddbt.lib.store;

import scala.Function1;
import scala.NotImplementedError;
import scala.Unit;


import java.util.ArrayList;

/**
 * Index interface.
 * By default it emits a warning when an unsupported function is called.
 * Specific implementation are below.
 *
 * @author TCK
 */
public abstract class Idx<E extends Entry> {
    static boolean allowResize = true;

    Idx(Store<E> st, int idx, boolean unique) {
        this.ops = st.ops()[idx];
        this.idx = idx;
        this.unique = unique;
    }

    protected Entry dataHead;
    protected final EntryIdx<E> ops;
    protected final int idx;
    protected final boolean unique;
    protected int size;

    protected void w(String n) {
        throw new NotImplementedError(this.getClass().getName() + ": " + n + " not supported");
    }

    public void unsafeInsert(E e) {
        w("unsafeInsert");
    }

    public String getSizeStat() {
        return "";
    }

    public void delete(E e) {
        w("delete");
    }

    public void deleteCopyDependent(E e) {
        delete(e);
    }

    public void deleteCopy(E e, Idx<E> primary) {
        E ref = primary.get(e);
        delete(ref);
    }

    public void update(E e) {
        w("update");
    } // reposition the entry if key/hash modified

    public void updateCopyDependent(E e, E ref) {
        delete(ref);
        unsafeInsert(e);
    }

    public void updateCopy(E e, Idx<E> primary) {
        E ref = primary.get(e);
        delete(ref);
        unsafeInsert(e);
    }

    public E get(E key) {
        w("get");
        return null;
    } // returns the first element only

    public E getCopy(E key) {
        E e = get(key);
        if (e != null)
            return (E) e.copy();
        return null;
    }

    public E getCopyDependent(E key) {
        E e = get(key);
        if (e != null)
            return (E) e.copy();
        return null;
    }
    public E getForUpdate(E key) {
        return get(key);
    }

    public void foreachResMap(Function1<E, Unit> f, MultiRes res) {
        //TODO :Fix later
        foreach(f);
    }

    public MultiRes foreachRes() {
        //TODO: Fix later
        return new ddbt.lib.store.ForEachRes<E>(null);
    }

    public void foreach(Function1<E, Unit> f) {
        w("foreach");
    } // on all elements; warning: what about reordering updates?

    public void foreachCopy(Function1<E, Unit> f) { w("foreachCopy"); }

    public void slice(E key, Function1<E, Unit> f) {
        w("slice");
    } // foreach on a slice

    public void sliceNoUpdate(E key, Function1<E, Unit> f) {
        w("sliceNoUpdate");
    }

    public MultiRes sliceRes(E key) {
        return null;
    }

    public void sliceResMap(E key, Function1<E, Unit> f, MultiRes res) {
        w("sliceResMap");
    }

    public void sliceResMapNoUpd(E key, Function1<E, Unit> f, MultiRes res) {
        w("sliceResMapNoUpd");
    }

    public void sliceCopy(E key, Function1<E, Unit> f) {
        w("sliceCopy");
    } // foreach on a slice

    public void sliceCopyDependent(E key, Function1<E, Unit> f) {
        sliceCopy(key, f);
    } // foreach on a slice

    public void range(E min, E max, boolean withMin, boolean withMax, Function1<E, Unit> f) {
        w("range");
    }

    public void clear() {
        w("clear");
    }

    public void compact() {
        w("compact");
    }

    public void getBucketStats() {
    }

    public boolean hashDiffers(E e1, E e2) {
        return ops.hash(e1) != ops.hash(e2);
    }

    public String info() {
        return this.getClass().getName() + "(" + idx + "," + (unique ? "unique" : "multiple") + ")";
    }
}

/**
 * IdxHash is a hashed index.
 * Buckets elements are stored in a list.
 * Use case: default, unless special operations are required.
 * <p>
 * Supported operations:
 * + insert,delete,update,get = O(1)
 * + range = O(n)
 */
class IdxHashEntry<E extends Entry> {
    IdxHashEntry(int h, E d) {
        hash = h;
        data = d;
    }

    int hash;
    E data;
    IdxHashEntry<E> next, prev;
}

@SuppressWarnings("unchecked")
class IdxHash<E extends Entry> extends Idx<E> {
    private static final int init_capacity = 16;
    private static final int max_capacity = 1 << 30;
    private static final float compact_factor = 0.05f;
    private final float load_factor;
    protected IdxHashEntry<E>[] data = new IdxHashEntry[init_capacity];
    protected int threshold;

    IdxHash(Store<E> st, int idx, boolean unique, float factor) {
        super(st, idx, unique);
        load_factor = factor;
        threshold = Math.round(init_capacity * load_factor);
    }

    IdxHash(Store<E> st, int idx, boolean unique) {
        super(st, idx, unique);
        load_factor = 0.75f;
        threshold = Math.round(init_capacity * load_factor);
    }

    @Override
    public String getSizeStat() {
        int numInArray = 0;
        for (int i = 0; i < data.length; ++i) {
            if (data[i] != null) {
                numInArray++;
            }
        }
        return "array length = " + data.length + "  numElements  = " + size + "  threshold = " + threshold + "   numInArray = " + numInArray;
    }

    // Private/inlined functions
    void _resize(int new_capacity) {
        IdxHashEntry<E>[] d = new IdxHashEntry[new_capacity];
        for (int i = 0, n = data.length; i < n; ++i) {
            IdxHashEntry<E> e = data[i], en;
            while (e != null) {
                en = e.next;
                int b = (e.hash > 0 ? e.hash : -e.hash) % new_capacity;
                e.next = d[b];
                if (d[b] != null)
                    d[b].prev = e;
                d[b] = e;
                e.prev = null;
                e = en;
            }
        }
        data = d;
        threshold = Math.min(Math.round(new_capacity * load_factor), max_capacity + 1);
    }

    protected void _resize() {
        int n = data.length;
        if (n == max_capacity) threshold = Integer.MAX_VALUE;
        else _resize(n << 1);
    }

    protected void _del(E e, IdxHashEntry<E> i) {
        if (idx == 0) {
            Entry elemPrev = e.prev, elemNext = e.next;
            if (elemPrev != null)
                elemPrev.next = elemNext;
            else
                dataHead = elemNext;

            if (elemNext != null) elemNext.prev = elemPrev;

            e.next = null;
            e.prev = null;
        }
        IdxHashEntry<E> nxt = i.next;
        IdxHashEntry<E> prv = i.prev;
        if (prv != null) { //not head
            prv.next = nxt;
            if (nxt != null)
                nxt.prev = prv;
        } else if (nxt != null) { //head and has other elements
            i.hash = nxt.hash;
            i.next = nxt.next;
            if (nxt.next != null)
                nxt.next.prev = i;
            i.data = nxt.data;
            nxt.data.data[idx] = i;
        } else { //head and the only element
            int h = i.hash, b = (h > 0 ? h : -h) % data.length;
            if (data[b] == i)
                data[b] = null;
            else
                throw new IllegalStateException();
        }
    }

    // Public
    @Override
    public void unsafeInsert(E e) {
        if (size == threshold) {
            _resize();
            if (!allowResize)
                throw new IllegalStateException("Heap resize");
        }
        int h = ops.hash(e), b = (h > 0 ? h : -h) % data.length;
        if (idx == 0) {
            e.prev = null;
            e.next = dataHead;
            if (dataHead != null)
                dataHead.prev = e;
            dataHead = e;
        }
        IdxHashEntry<E> i = new IdxHashEntry<E>(h, e);
        e.data[idx] = i;
        i.next = data[b];
        if (data[b] != null)
            data[b].prev = i;
        data[b] = i;
        size += 1;
    }

    @Override
    public void getBucketStats() {
        int maxEntries = 0;
        int maxSlices = 0;
        int numBuckets = 0;
        int numSlices = 0;
        int numEntries = 0;

        for (int b = 0; b < data.length; ++b) {
            IdxHashEntry<E> n1 = data[b];
            if (n1 == null)
                continue;
            numBuckets++;
            int ns = 0;
            int es = 0;
            IdxHashEntry<E> n2 = n1;
            do {
                do {
                    ++es;
                } while ((n2 = n2.next) != null && n2.hash == n1.hash && ops.cmp(n1.data, n2.data) == 0);
                ++ns;
            } while ((n1 = n2) != null);
            if (es > maxEntries)
                maxEntries = es;
            if (ns > maxSlices)
                maxSlices = ns;
            numEntries += es;
            numSlices += ns;
        }
        if (numBuckets == 0)
            System.err.println("Empty");
        else {
            String idxs = "IDX = " + idx;
            String ents = "    Entries : total = " + numEntries + "  avg = " + (numEntries / (1.0 * numBuckets)) + "  max = " + maxEntries;
            String slices = "    Slices : total = " + numBuckets + "  avg = " + (numSlices / (1.0 * numBuckets)) + "  max = " + maxSlices;
            System.err.println(idxs + ents + slices);
        }
    }


    @Override
    public void delete(E e) {
        IdxHashEntry<E> i = (IdxHashEntry<E>) e.data[idx];
        if (i != null) {
            _del(e, i);
            e.data[idx] = null;
            size -= 1;
        }
    }

    @Override
    public void update(E e) {
        IdxHashEntry<E> i = (IdxHashEntry<E>) e.data[idx];
        if (i != null && i.hash != ops.hash(e)) {
            _del(e, i);
            size -= 1;
            unsafeInsert(e);
        }
    }

    @Override
    public E get(E key) {
        int h = ops.hash(key);
        IdxHashEntry<E> e = data[(h > 0 ? h : -h) % data.length];
        while (e != null && (e.hash != h || ops.cmp(key, e.data) != 0))
            e = e.next;
        return e != null ? e.data : null;
    }

    @Override
    public void foreach(Function1<E, Unit> f) {
        Entry cur = dataHead;
        while (cur != null) {
            f.apply((E) cur);
            cur = cur.next;
        }
    }

    @Override
    public void foreachCopy(Function1<E, Unit> f) {
        ArrayList<E> entries = new ArrayList<E>();
        Entry cur = dataHead;
        while (cur != null) {
            entries.add((E) cur.copy());
            cur = cur.next;
        }
        for(E e_ : entries)
            f.apply(e_);
    }

    @Override
    public MultiRes sliceRes(E key) {
        if (idx == 0) {
            //Can't use primary hash fn for slice
            throw new NotImplementedError("slice res not implemented for primary idx");
        } else {
            int h = ops.hash(key);
            IdxHashEntry<E> e = data[(h > 0 ? h : -h) % data.length];
            if (e != null) do {
                if (e.hash == h && ops.cmp(key, e.data) == 0)
                    return new SliceRes<E>(e);
                e = e.next;
            } while (e != null);
            return new SliceRes<E>(null);
        }
    }

    @Override
    public void sliceResMapNoUpd(E key, Function1<E, Unit> f, MultiRes res) {
        if (idx == 0) {
            //Can't use primary hash fn for slice
            throw new NotImplementedError("sliceresMapNoUpd not implemented for primary idx");
        } else {
            IdxHashEntry<E> e = ((SliceRes<E>) res).sliceHead();
            int h = e.hash;
            do {
                if (e.hash == h && ops.cmp(key, e.data) == 0)
                    f.apply(e.data);
            } while ((e = e.next) != null);
        }
    }

    @Override
    public void sliceResMap(E key, Function1<E, Unit> f, MultiRes res) {
        if (idx == 0) {
            //Can't use primary hash fn for slice
            throw new NotImplementedError("sliceResMap not implemented for primary idx");
        } else {
            IdxHashEntry<E> e = ((SliceRes<E>) res).sliceHead();
            ArrayList<E> entries = new ArrayList<E>();
            int h = e.hash;
            do {
                if (e.hash == h && ops.cmp(key, e.data) == 0)
                    entries.add(e.data);
            } while ((e = e.next) != null);
            for (E e_ : entries) {
                f.apply(e_);
            }
        }
    }

    @Override
    public void sliceNoUpdate(E key, Function1<E, Unit> f) {
        if (idx == 0) {
            //Can't use primary hash fn for slice
            Entry e = dataHead;
            while(e != null) {
                if(ops.cmp(key, (E)e) == 0)
                    f.apply((E)e);
                e = e.next;
            }
        } else {
            int h = ops.hash(key);
            IdxHashEntry<E> e = data[(h > 0 ? h : -h) % data.length];
            if (e != null) do {
                if (e.hash == h && ops.cmp(key, e.data) == 0) f.apply(e.data);
                e = e.next;
            } while (e != null);
        }
    }

    @Override
    public void slice(E key, Function1<E, Unit> f) {
        ArrayList<E> entries = new ArrayList<E>();
        if (idx == 0) {
            //Can't use primary hash fn for slice
            Entry e = dataHead;
            while(e != null) {
                if(ops.cmp(key, (E)e) == 0)
                    entries.add((E) e);
                e = e.next;
            }
        } else {
            int h = ops.hash(key);
            IdxHashEntry<E> e = data[(h > 0 ? h : -h) % data.length];
            if (e != null) do {
                if (e.hash == h && ops.cmp(key, e.data) == 0) entries.add(e.data);
                e = e.next;
            } while (e != null);
        }
        for (E e_ : entries) {
            f.apply(e_);
        }
    }

    @Override
    public void sliceCopy(E key, Function1<E, Unit> f) {
        ArrayList<E> entries = new ArrayList<E>();
        if (idx == 0) {
            //Can't use primary hash fn for slice
            Entry e = dataHead;
            while(e != null) {
                if(ops.cmp(key, (E)e) == 0)
                    entries.add((E) e.copy());
                e = e.next;
            }
        } else {
            int h = ops.hash(key);
            IdxHashEntry<E> e = data[(h > 0 ? h : -h) % data.length];
            if (e != null) do {
                if (e.hash == h && ops.cmp(key, e.data) == 0)
                    entries.add((E) e.data.copy());
                e = e.next;
            } while (e != null);
        }
        for (E e_ : entries) {
            f.apply(e_);
        }
    }

    @Override
    public void range(E min, E max, boolean withMin, boolean withMax, Function1<E, Unit> f) {
        int cMin = withMin ? -1 : 0;
        int cMax = withMax ? 1 : 0;
        for (int i = 0, n = data.length; i < n; ++i) {
            IdxHashEntry<E> e = data[i], em, en;
            while (e != null) {
                en = e.next;
                if (ops.cmp(e.data, min) > cMin && ops.cmp(e.data, max) < cMax) f.apply(e.data);
                e = en;
            }
        }
    }

    @Override
    public void clear() {
        IdxHashEntry<E> z = null;
        dataHead = null;
        for (int i = 0, n = data.length; i < n; ++i) {
            data[i] = z;
        }
        size = 0;
    }

    @Override
    public void compact() {
        if (data.length * compact_factor > size)
            _resize(Math.max(init_capacity, 1 << ((int) Math.ceil(1 + (Math.log((size / load_factor)) / Math.log(2))))));
    }

    @Override
    public String info() {
        IdxHashEntry<E> e, en;
        int max = 0;
        int n = data.length;
        for (int i = 0; i < n; ++i) {
            e = data[i];
            int c = 0;
            if (e != null) do {
                en = e.next;
                ++c;
                e = en;
            } while (e != null);
            if (c > max) max = c;
        }
        return "IdxHash(" + idx + "," + (unique ? "unique" : "multiple") + ") {\n  elements => " + size + "\n  buckets => " + n + "\n  occupancy => " + (size * 1.0 / n) + "\n  occupancyMax => " + max + "\n  loadFactor => " + load_factor + "\n}";
    }
}

/**
 * IdxSliced maintains 1 element (min or max) for each slice.
 * IdxSliced maintains minimum or maximum of each slice. Slicing is provided by
 * sliceIdx which correspond to another _implemented_ index supporting slice op.
 * When min/max is deleted, a new min/max is looked up in the sliceIdx index.
 * <p>
 * Use case: min/max is rarely deleted.
 * <p>
 * Supported operations:
 * + insert,delete,update = O(1)/O(N) if min/max / get=O(1)
 * + clear,size
 */
@SuppressWarnings("unchecked")
class IdxSliced<E extends Entry> extends IdxHash<E> {
    private int cmpRes;
    private int cmpIdx;
    private E cmpE = null;
    private Idx<E>[] idxs;
    private EntryIdx<E> ops2;

    IdxSliced(Store<E> st, int cmpIdx, int sliceIdx, boolean max) {
        super(st, sliceIdx, true);
        ops2 = st.ops()[cmpIdx];
        this.idxs = st.idxs();
        cmpIdx = cmpIdx;
        cmpRes = max ? 1 : -1;
    }

    @Override
    //SBJ: TODO: Remove checks
    public void unsafeInsert(E e) {
        if (size == threshold) _resize();
        int h = ops.hash(e), b = (h > 0 ? h : -h) % data.length;
        IdxHashEntry<E> p = data[b], i = new IdxHashEntry<E>(h, e);
        e.data[idx] = i;
        if (p == null) {
            data[b] = i;
            size += 1;
            return;
        } else do {
            if (p.hash == h && ops.cmp(e, p.data) == 0) {
                if (ops2.cmp(e, p.data) == cmpRes) {
                    p.data = e;
                    e.data[idx] = p;
                }
                return;
            }
            if (p.next == null) {
                p.next = i;
                size += 1;
                return;
            }
            p = p.next;
        } while (p != null);
    }

    @Override
    public void delete(E e) {
        IdxHashEntry<E> i = (IdxHashEntry<E>) e.data[idx];
        if (i == null) return;
        _del(e, i);
        e.data[idx] = null;
        // Find a replacement candidate
        Function1<E, Unit> f = new JFun1<E, Unit>() {
            @Override
            public Unit apply(E e) {
                if (cmpE == null || ops.cmp(e, cmpE) == cmpRes) cmpE = e;
                return (Unit) null;
            }
        };
        idxs[idx].slice(e, f);
        if (cmpE != null) unsafeInsert(cmpE);
        else size -= 1;
        cmpE = null;
    }
}

/**
 * IdxList is a simple sequential storage reminiscent of the disk era.
 * Lightweight singe-linked list (head --> tail), unique should be disabled
 * for best insertion performance (avoid checking duplicates).
 * <p>
 * Use case: WAL log or temporary storage, sorted by least recently inserted.
 * <p>
 * Supported operations:
 * + insert O(1)/O(N),delete O(N),update O(1)/O(N) / get=O(1)
 * + clear,size
 */
@SuppressWarnings("unchecked")
class IdxList<E extends Entry> extends Idx<E> {
    IdxList(Store<E> st, int idx, boolean unique) {
        super(st, idx, unique);
    }

    private E head = null;
    private E tail = null;

    @Override
    //SBJ: TODO: Remove checks
    public void unsafeInsert(E e) {
        /*
        if (unique && head != null) {
            if (e == head) {
                head = (E) head.data[idx];
                if (e == tail) tail = null;
            } else {
                E p = head;
                do {
                    E n = (E) p.data[idx];
                    if (e == n) {
                        p.data[idx] = n.data[idx];
                        if (n == tail) tail = p;
                        break;
                    }
                    p = n;
                } while (p != null);
            }
        }
        */
        if (tail != null) {
            tail.data[idx] = e;
            tail = e;
        } else {
            head = e;
            tail = e;
        }
    }

    @Override
    public void delete(E e) {
        if (head == null) return;
        if (e == head) {
            head = (E) head.data[idx];
            if (e == tail) tail = null;
        } else {
            E p = head;
            do {
                E n = (E) p.data[idx];
                if (e == n) {
                    p.data[idx] = n.data[idx];
                    if (n == tail) tail = p;
                    break;
                } else p = n;
            } while (p != null);
        }
        e.data[idx] = null;
    }

    @Override
    public void update(E e) {
        delete(e);
        unsafeInsert(e);
    }

    @Override
    public E get(E key) {
        E p = head;
        if (p != null) do {
            if (ops.cmp(key, p) == 0) return p;
            p = (E) p.data[idx];
        } while (p != null);
        return null;
    }

    @Override
    public void foreach(Function1<E, Unit> f) {
        E p = head;
        if (p != null) do {
            E n = (E) p.data[idx];
            f.apply(p);
            p = n;
        } while (p != null);
    }
    @Override
    public void foreachCopy(Function1<E, Unit> f) {
        E p = head;
        ArrayList<E> entries = new ArrayList<E>();
        if (p != null) do {
            E n = (E) p.data[idx];
            entries.add((E) p.copy());
            p = n;
        } while (p != null);
        for(E e_ : entries)
            f.apply(e_);
    }
    @Override
    public void slice(E key, Function1<E, Unit> f) {
        E p = head;
        if (p != null) do {
            E n = (E) p.data[idx];
            if (ops.cmp(key, p) == 0) f.apply(p);
            p = n;
        } while (p != null);
    }

    @Override
    public void sliceCopy(E key, Function1<E, Unit> f) {
        E p = head;
        ArrayList<E> entries = new ArrayList<E>();
        if (p != null) do {
            E n = (E) p.data[idx];
            if (ops.cmp(key, p) == 0) entries.add((E) p.copy());
            p = n;
        } while (p != null);
        for (E e_ : entries)
            f.apply(e_);
    }

    @Override
    public void clear() {
        head = null;
        tail = null;
    }

    @Override
    public void compact() {
    } // nothing to do
}

File: storelib/src/lib/storeScala/Idx.scala
package ddbt.lib.storeScala

import java.util

abstract class Idx[E <: Entry] {
  protected var dataHead: Entry = null
  protected var ops: EntryIdx[E] = null
  protected var idx = 0
  protected var unique = false
  var size = 0//GS investigate why it cannot be protected

  def this(st: Store[E], idx: Int, unique: Boolean) {
    this()
    this.ops = st.ops(idx)
    this.idx = idx
    this.unique = unique
  }

  def w(n: String) {
    throw new NotImplementedError(this.getClass().getName() + ": " + n + " not supported");
  }

  def unsafeInsert(e: E): Unit = {
    w("unsafeInsert")
  }

  def getSizeStat = ""

  def delete(e: E): Unit = {
    w("delete")
  }

  def deleteCopy(e: E, primary: Idx[E]) {
    val ref = primary.get(e)
    delete(ref)
  }

  def deleteCopyDependent(e: E): Unit = {
    delete(e)
  }

  def update(e: E): Unit = {
    w("update")
    // reposition the entry if key/hash modified}
  }

  def updateCopy(e: E, primary: Idx[E]) {
    val ref = primary.get(e);
    delete(ref);
    unsafeInsert(e);
  }

  def updateCopyDependent(e: E, ref: E): Unit = {
    delete(ref)
    unsafeInsert(e)
  }

  def get(key: E): E = {
    w("get")
    null.asInstanceOf[E]
  }

  def getCopy(key: E): E = {
    val e = get(key)
    (if (e != null) e.copy else null).asInstanceOf[E]

  }

  def getCopyDependent(key: E): E = {
    val e = get(key)
    (if (e != null) e.copy else null).asInstanceOf[E]

  }

  def foreach(f: Function1[E, Unit]) = w {
    "foreach"
  }

  def foreachCopy(f: Function1[E, Unit]): Unit = {
    w("foreachCopy")
  }

  def foreachRes(): MultiRes = {
    //TODO: Fix later
    new ForEachRes[E](null.asInstanceOf[E])
  }

  def foreachResMap(f: Function1[E, Unit], res: MultiRes): Unit = {
    //TODO :Fix later
    foreach(f)
  }

  def slice(key: E, f: Function1[E, Unit]) = w("slice")

  def sliceCopy(key: E, f: Function1[E, Unit]) = w("sliceCopy")

  def sliceCopyDependent(key: E, f: Function1[E, Unit]) = sliceCopy(key, f)

  def sliceRes(key: E): MultiRes = {
    w("sliceRes"); null
  }

  def sliceResMapNoUpd(key: E, f: Function1[E, Unit], res: MultiRes): Unit = {
    w("sliceResMapNoUpd")
  }

  def range(min: E, max: E, withMin: Boolean, withMax: Boolean, f: Function1[E, Unit]): Unit = {
    w("range")
  }

  def clear(): Unit = {
    w("clear")
  }

  def compact(): Unit = {
    w("compact")
  }
  def hash(e1:E)=ops.hash(e1)
  def hashDiffers(e1: E, e2: E): Boolean = ops.hash(e1) != ops.hash(e2)

  def info: String = this.getClass.getName + "(" + idx + "," + (if (unique) "unique" else "multiple") + ")"
}

class IdxHashEntry[E <: Entry](var hash: Int, var data: E, var next: IdxHashEntry[E] = null, var prev: IdxHashEntry[E] = null) {
}

class IdxHash[E <: Entry](st: Store[E], idx: Int, unique: Boolean) extends Idx[E](st, idx, unique) {
  //TODO: Check if this would work
  private val init_capacity = 16
  private val max_capacity = 1 << 30
  private val compact_factor = 0.05f
  private var load_factor = 0.75f
  protected var data = new Array[IdxHashEntry[E]](init_capacity)
  protected var threshold = init_capacity * load_factor.round
  private val allowResize = true

  def _resize(): Unit = {
    //GS investigate why it cannot be protected
    val n = data.length
    if (n == max_capacity) threshold = Integer.MAX_VALUE
    else _resize(n * 10)
  }

  def _resize(new_capacity: Int): Unit = {
    //GS investigate why it cannot be protected
    val d = new Array[IdxHashEntry[E]](new_capacity)
    var i = 0

    while (i < data.length) {
      var e = data(i)
      var en = null.asInstanceOf[IdxHashEntry[E]]
      while (e != null) {
        en = e.next
        val b = (if (e.hash > 0) e.hash
        else -e.hash) % new_capacity
        e.next = d(b)
        if (d(b) != null) d(b).prev = e
        d(b) = e
        e.prev = null
        e = en
      }
      i += 1
    }
    data = d
    threshold = Math.min(new_capacity * load_factor.round, max_capacity + 1)
  }

  override def unsafeInsert(e: E): Unit = {
    if (size == threshold) {
      _resize()
      if (!allowResize) throw new IllegalStateException("Heap resize")
    }
    val h = ops.hash(e)
    val b = (if (h > 0) h else -h) % data.length
    if (idx == 0) {
      e.prev = null
      e.next = dataHead
      if (dataHead != null) dataHead.prev = e
      dataHead = e
    }

    val i = new IdxHashEntry[E](h, e)
    e.data(idx) = i
    i.next = data(b)
    if (data(b) != null) data(b).prev = i
    data(b) = i
    size += 1
  }

  protected def _del(e: E, i: IdxHashEntry[E]): Unit = {
    if (idx == 0) {
      val elemPrev = e.prev
      val elemNext = e.next
      if (elemPrev != null) elemPrev.next = elemNext
      else dataHead = elemNext
      if (elemNext != null) elemNext.prev = elemPrev
      e.next = null
      e.prev = null
    }
    val nxt = i.next
    val prv = i.prev
    if (prv != null) { //not head
      prv.next = nxt
      if (nxt != null) nxt.prev = prv
    }
    else if (nxt != null) { //head and has other elements
      i.hash = nxt.hash
      i.next = nxt.next
      if (nxt.next != null) nxt.next.prev = i
      i.data = nxt.data
      nxt.data.data(idx) = i
    }
    else { //head and the only element
      val h = i.hash
      val b = (if (h > 0) h
      else -h) % data.length
      if (data(b) eq i) data(b) = null
      else throw new IllegalStateException
    }
  }
  override def delete(e: E): Unit = {
    val i = e.data(idx).asInstanceOf[IdxHashEntry[E]]
    if (i != null) {
      _del(e, i)
      e.data(idx) = null
      size -= 1
    }
  }
  override def get(key: E): E = {
    val h = ops.hash(key)
    var e = data((if (h > 0) h else -h) % data.length)
    while (e != null && (e.hash != h || (ops.cmp(key, e.data) != 0))) e = e.next //TODO:don't know what's wrong here
    (if (e != null) e.data else null).asInstanceOf[E]
  }

  override def foreach(f: Function1[E, Unit]): Unit = {
    var cur = dataHead
    while (cur != null) {
      f.apply(cur.asInstanceOf[E])
      cur = cur.next
    }
  }

  override def slice(key: E, f: Function1[E, Unit]): Unit = {
    //val entries = collection.mutable.ListBuffer[E]()//Apparently ArrayList exists in Scala.js as well, but I was having trouble with it (But also set isn't working)
    val entries = new util.ArrayList[E]()
    if (idx == 0) { //Can't use primary hash fn for slice
      var e = dataHead
      while (e != null) {
        if (ops.cmp(key, e.asInstanceOf[E]) == 0) entries.add(e.asInstanceOf[E])
        e = e.next
      }
    }
    else {
      val h = ops.hash(key)
      var e = data((if (h > 0) h else -h) % data.length)
      if (e != null) do {
        if (e.hash == h && (ops.cmp(key, e.data) == 0)) entries.add(e.data)
        e = e.next
      } while (e != null)
    }
    //    for (e_ <- entries) {
    //      f.apply(e_)
    //    }
    while (entries.iterator().hasNext) {
      f.apply(entries.iterator().next())
    }
  }


  override def sliceRes(key: E): MultiRes =
    if (idx == 0) { //Can't use primary hash fn for slice
      throw new NotImplementedError("slice res not implemented for primary idx")
    }
    else {
      val h = ops.hash(key)
      var e = data((if (h > 0) h else -h) % data.length)
      if (e != null)
        do {
          if ((e.hash == h) && (ops.cmp(key, e.data) == 0)) return new SliceRes[E](e)
          e = e.next
        } while (e != null)
      new SliceRes[E](null)
    }


  override def sliceResMapNoUpd(key: E, f: Function1[E, Unit], res: MultiRes): Unit = {
    if (idx == 0) { //Can't use primary hash fn for slice
      throw new NotImplementedError("sliceresMapNoUpd not implemented for primary idx")
    }
    else {
      var e = res.asInstanceOf[SliceRes[E]].sliceHead
      val h = e.hash
      do {
        if ((e.hash == h) && (ops.cmp(key, e.data) == 0)) f.apply(e.data)
        e = e.next
      } while (e != null)
    }
  }

  override def clear(): Unit = {
    val z = null
    dataHead = null
    var i = 0
    val n = data.length
    while (i < n) {
      data(i) = z
      i += 1
    }
    size = 0
  }

}
